### PR TITLE
Node execution refactor for Parent-child relationship

### PR DESCRIPTION
**Pull request by anandswaminathan**
Addressed based on discussion from https://github.com/lyft/flytepropeller/pull/174

Intentions:
The code if rolled out as is - will have no impact or changes to event schema. The code needs to be rolled out and the event version has to be bumped up to V1 for the changes to reflect both in the CRD and the Admin.

Details:
This refactor mainly touches the way node execution status for child nodes are stored in etcd and details relayed to Admin. Following new fields have been introduced :

UniqueNodeID - Indicates the unique execution of the node within the workflow. Func(ParentUniqueID, currentNodeId, attempt)
ParentAttempts - Execution attempt when the Parent is retried as whole.
WorkflowMeta - EventVersion: Event Version for a giver workflow.
In V2,

@EngHabu @kumare3

### CODE REVIEW

File: pkg/apis/flyteworkflow/v1alpha1/nodes.go [hunk 1]
```diff
@@ -94,6 +94,7 @@ func (in *NodeMetadata) DeepCopyInto(out *NodeMetadata) {
 
 type NodeSpec struct {
 	ID            NodeID                        `json:"id"`
+	Name          string                        `json:"name,omitempty"`
```
Comment (kumare3) [inline pkg/apis/flyteworkflow/v1alpha1/nodes.go:97]: who populates this?

File: pkg/controller/nodes/common/utils.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+	if parentInfo != nil {
+		parentUniqueId = parentInfo.GetUniqueID()
+		parentRetryAttempt = strconv.Itoa(int(parentInfo.CurrentAttempt()))
+	}
+
+	return utils.FixedLengthUniqueIDForParts(20, parentUniqueId, parentRetryAttempt, nodeID)
+}
+
+func GetParentInfo(grandParentInfo executors.ImmutableParentInfo, nodeID string, parentAttempt uint32) (executors.ImmutableParentInfo, error){
```
Comment (kumare3) [inline pkg/controller/nodes/common/utils.go:22]:
```suggestion
func CreateParentInfo(grandParentInfo executors.ImmutableParentInfo, nodeID string, parentAttempt uint32) (executors.ImmutableParentInfo, error){
```

File: pkg/controller/nodes/common/utils.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package common
+
+import (
+	"github.com/lyft/flytepropeller/pkg/apis/flyteworkflow/v1alpha1"
+	"github.com/lyft/flytepropeller/pkg/controller/executors"
+	"github.com/lyft/flytepropeller/pkg/utils"
+	"strconv"
+)
+
+func GenerateUniqueId(parentInfo executors.ImmutableParentInfo, nodeID string) (string, error) {
```
Comment (kumare3) [inline pkg/controller/nodes/common/utils.go:10]: lets add a comment explaining what this is

File: pkg/controller/nodes/common/utils.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	var parentUniqueId v1alpha1.NodeID
+	var parentRetryAttempt string
+
+	if parentInfo != nil {
+		parentUniqueId = parentInfo.GetUniqueID()
+		parentRetryAttempt = strconv.Itoa(int(parentInfo.CurrentAttempt()))
+	}
+
+	return utils.FixedLengthUniqueIDForParts(20, parentUniqueId, parentRetryAttempt, nodeID)
```
Comment (kumare3) [inline pkg/controller/nodes/common/utils.go:19]: why 20 chars, it is fine, but, maybe make this a constant?

File: pkg/controller/nodes/executor.go [hunk 5]
```diff
@@ -792,8 +797,20 @@ func (c *nodeExecutor) AbortHandler(ctx context.Context, execContext executors.E
...
+		}
+		if nCtx.ExecutionContext().GetEventVersion() != v1alpha1.V0 {
+			currentNodeUniqueID, err := common.GenerateUniqueId(nCtx.ExecutionContext().GetParentInfo(), nodeExecutionId.NodeId)
+			if err != nil {
+				return err
+			}
+			nodeExecutionId.NodeId = currentNodeUniqueID
+		}
+
 		err = c.IdempotentRecordEvent(ctx, &event.NodeExecutionEvent{
```
Comment (kumare3) [inline pkg/controller/nodes/executor.go:812]: do we not need to pass the node name as well?

File: pkg/controller/nodes/subworkflow/subworkflow.go [hunk 6]
```diff
@@ -56,8 +60,14 @@ func (s *subworkflowHandler) startAndHandleSubWorkflow(ctx context.Context, nCtx
 
 // Calls the recursive node executor to handle the SubWorkflow and translates the results after the success
 func (s *subworkflowHandler) handleSubWorkflow(ctx context.Context, nCtx handler.NodeExecutionContext, subworkflow v1alpha1.ExecutableSubWorkflow, nl executors.NodeLookup) (handler.Transition, error) {
-
-	state, err := s.nodeExecutor.RecursiveNodeHandler(ctx, nCtx.ExecutionContext(), subworkflow, nl, subworkflow.StartNode())
+	// The current node would end up becoming the parent for the sun workflow nodes.
```
Comment (kumare3) [inline pkg/controller/nodes/subworkflow/subworkflow.go:63]:
```suggestion
	// The current node would end up becoming the parent for the sub-workflow nodes.
```

File: pkg/controller/nodes/transformers.go [hunk 7]
```diff
@@ -82,13 +91,28 @@ func ToNodeExecutionEvent(nodeExecID *core.NodeExecutionIdentifier, info handler
...
 
-	// TODO this should use node-node relationship instead of taskID
-	if status.GetParentTaskID() != nil {
+	if eventVersion == v1alpha1.V0 && status.GetParentTaskID() != nil {
 		nev.ParentTaskMetadata = &event.ParentTaskExecutionMetadata{
 			Id: status.GetParentTaskID(),
 		}
 	}
 
+	if eventVersion != v1alpha1.V0 {
```
Comment (kumare3) [inline pkg/controller/nodes/transformers.go:100]: I know it does not exist, but can you add a few unit tests for this code?

**Comment by kumare3** [comment]
@anandswaminathan this looks awesome, you can actually test it end to end locally, thats what I do. setup a Sandbox, and delete propeller, use propeller in your IDE

File: pkg/apis/flyteworkflow/v1alpha1/nodes.go (hunk 1 above)
Comment (anandswaminathan) [reply pkg/apis/flyteworkflow/v1alpha1/nodes.go:97]: Compiler - BuildFlyteWorkflow

File: pkg/controller/nodes/common/utils.go (hunk 3 above)
Comment (anandswaminathan) [reply pkg/controller/nodes/common/utils.go:10]: YES

File: pkg/controller/nodes/common/utils.go (hunk 4 above)
Comment (anandswaminathan) [reply pkg/controller/nodes/common/utils.go:19]: Sure.

File: pkg/controller/nodes/executor.go (hunk 5 above)
Comment (anandswaminathan) [reply pkg/controller/nodes/executor.go:812]: Nope. this is abort. So only status is updated.

File: pkg/controller/nodes/transformers.go (hunk 7 above)
Comment (anandswaminathan) [reply pkg/controller/nodes/transformers.go:100]: No unit tests have been added. I am on it.

File: pkg/compiler/transformers/k8s/node.go [hunk 8]
```diff
@@ -58,13 +58,18 @@ func buildNodeSpec(n *core.Node, tasks []*core.CompiledTask, errs errors.Compile
...
-	if n.GetMetadata() != nil && n.GetMetadata().GetInterruptibleValue() != nil {
-		interruptVal := n.GetMetadata().GetInterruptible()
-		interruptible = &interruptVal
+	var name string
+	if n.GetMetadata() != nil {
+		if n.GetMetadata().GetInterruptibleValue() != nil {
+			interruptVal := n.GetMetadata().GetInterruptible()
+			interruptible = &interruptVal
+		}
+		name = n.GetMetadata().Name
```
[the remark this answers is not in the document]
Comment (anandswaminathan) [reply pkg/compiler/transformers/k8s/node.go:67]: @EngHabu This is not needed, as name is only sent in version v2.

File: pkg/apis/flyteworkflow/v1alpha1/workflow.go [hunk 9]
```diff
@@ -49,6 +50,24 @@ type FlyteWorkflow struct {
...
+}
+
+type WorkflowMeta struct {
+	EventVersion EventVersion `json:"eventVersion,omitempty"`
+}
+
+type EventVersion int
+
+const (
+	V0 EventVersion = iota
```
[the remark this answers is not in the document]
Comment (anandswaminathan) [reply pkg/apis/flyteworkflow/v1alpha1/workflow.go:73]: Sure

File: pkg/apis/flyteworkflow/v1alpha1/workflow.go [hunk 10]
```diff
@@ -49,6 +50,24 @@ type FlyteWorkflow struct {
...
+	V1
```
[the remark this answers is not in the document]
Comment (anandswaminathan) [reply pkg/apis/flyteworkflow/v1alpha1/workflow.go]: done

File: pkg/controller/nodes/common/utils.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
+package common
```
Comment (kumare3) [inline pkg/controller/nodes/common/utils.go:1]: unit tests?

File: pkg/controller/nodes/executor.go (hunk 5 above)
Comment (kumare3) [reply pkg/controller/nodes/executor.go:813]: but still should we not be consistent?

File: pkg/controller/nodes/common/utils.go (hunk 11 above)
Comment (anandswaminathan) [reply pkg/controller/nodes/common/utils.go:1]: Sure. It is covered by few tests. Will add one for this as well.

**Comment by kumare3** [comment]
This looks great to me. This is a much needed change and would help us greatly!

**Comment by kumare3** [comment]
Lint failed?

File: pkg/controller/nodes/executor.go (hunk 5 above)
Comment (anandswaminathan) [reply pkg/controller/nodes/executor.go:813]: @kumare3 We do not populate immutable fields here. I am trying to be consistent here and think it's right way.

**Pull request merged**